src/main/kotlin/org/rust/lang/core/completion/LookupElements.kt

```diff
@@ -126,8 +130,19 @@ fun createLookupElement(element: RsCompositeElement, scopeName: String): LookupE
 fun LookupElementBuilder.withPriority(priority: Double): LookupElement =
     PrioritizedLookupElement.withPriority(this, priority)
 
-private val InsertionContext.getUseItem: RsUseItem?
-    get() = PsiTreeUtil.findElementOfClassAtOffset(file, tailOffset - 1, RsUseItem::class.java, false)
+private fun appendSemicolon(context: InsertionContext) {
+    val curUseItem = context.getItemOfType<RsUseItem>()
+    if (curUseItem != null) {
+        val hasSemicolon = curUseItem.lastChild!!.elementType == RsElementTypes.SEMICOLON
+        if (!(hasSemicolon || context.isInGlob)) {
+            context.addSuffix(";")
+        }
+    }
+}
```

Comment:
There's a bit of code duplication between this method and `is RsFunction` branch. Could we somehow find out, how to get rid of it? 

One possible solution I have in mind is to attach a single insertion handler to all our elements, and *inside* that handler determine the context by `tailOffset` and the element to be inserted via a `when` switch. That is, leave only a single `.withInsertionHandler` in this file, and move all inertion handling logic to a separate file. 

What do you think about this?

Reply:
The reason I left the duplication for functions is because they need the return of `@handler` in order to work.
But I think I see the idea behind your proposed refactoring and I like it. Although I am not quite sure how we can decouple `.withInsertionHandler` from the types while still having the different type- and tail-texts.

Reply:
>Although I am not quite sure how we can decouple .withInsertionHandler from the types while still having the different type- and tail-texts.

Hm, I think we need some function which switch over element type and returns an `LookupElementBuilder` with type text, tail text, but without insertion handler. 

Then, we add an insertion handler to that Builder

Reply:
Ok, I'll r+ this now, but it'd be great to remove duplication in a separate PR!